run/bitlocker2john.py

```diff
@@ -0,0 +1,237 @@
+FVE_ENTRY_TYPES = {0x0: 'None', 0x2: 'VMK', 0x3: 'FVEK', 0x4: 'Validation', 0x6: 'Startup key', 0x7: 'Computer description', 0xb: 'FVEK backup', 0xf: 'Volume header block'}
+FVE_VALUE_TYPES = {0x0: 'Erased', 0x1: 'Key', 0x2: 'UTF-16 string', 0x3: 'Stretch key', 0x4: 'Use key', 0x5: 'AES-CCM encrypted key', 0x6: 'TPM encoded key', 0x7: 'Validation', 0x8: 'VMK', 0x9: 'External key', 0xa: 'Update', 0xb: 'Error', 0xf: 'Offset and size'}
+ITERATION_COUNT = 0x100000
+BITLOCKER_PASSWORD_HASH_VERSIONS = [0,1] # 0,1 both supported on hashcat
+BITLOCKER_RECOVERY_HASH_VERSIONS = [2,3]
+HASHCAT_HASH = []
+
+def guid_to_hex(guid):
+    guid_parts = guid.split('-')
+
+    search_target = ''.join([guid_parts[0][i:i+2] for i in range(0, len(guid_parts[0]), 2)][::-1])
+    search_target += ''.join([guid_parts[1][i:i+2] for i in range(0, len(guid_parts[1]), 2)][::-1])
+    search_target += ''.join([guid_parts[2][i:i+2] for i in range(0, len(guid_parts[2]), 2)][::-1])
+    search_target += guid_parts[3]
+    search_target += guid_parts[4]
+
+    return search_target
+
+def hex_to_guid(hex_str):
+
+    guid_parts = [0] * 5
+    guid_parts[0] = hex_str[0:8]
+    guid_parts[1] = hex_str[8:12]
+    guid_parts[2] = hex_str[12:16]
+    guid_parts[3] = hex_str[16:20]
+    guid_parts[4] = hex_str[20:]
+
+    guid = ''.join([guid_parts[0][i:i+2] for i in range(0, len(guid_parts[0]), 2)][::-1])
+    guid += '-'
+    guid += ''.join([guid_parts[1][i:i+2] for i in range(0, len(guid_parts[1]), 2)][::-1])
+    guid += '-'
+    guid += ''.join([guid_parts[2][i:i+2] for i in range(0, len(guid_parts[2]), 2)][::-1])
+    guid += '-'
+    guid += guid_parts[3]
+    guid += '-'
+    guid += guid_parts[4]
+
+    return guid.upper()
+
+def uint_to_int(b):
+    return int(b[::-1].hex(), 16)
+
+def parse_FVEK(fvek_data):
+    print("\nParsing FVEK...")
+    nonce    = fvek_data[:12]
+    mac      = fvek_data[12:28]
+    enc_data = fvek_data[28:]
+
+    print("Mac:", mac.hex())
+    print("Nonce:", nonce.hex())
+    print("Encrypted data:", enc_data.hex())
+
+    return nonce, mac, enc_data
+
+def parse_stretch_key(data):
+    print("\nParsing stretch key...")
+    encryption_method = hex(uint_to_int(data[0:4]))
+    salt = data[4:20]
+    print("Encryption method:", encryption_method)
+    print("Salt:", salt.hex())
+    current_pos = 0
+    aes_ccm_data = data[20:]
+    current_pos, data, value_type = parse_fve_metadata_entry(current_pos, aes_ccm_data)
+    nonce, mac, enc_data = parse_aes_ccm_encrypted_key(data)
+
+    return salt, nonce, mac, enc_data
+
+def generate_hashcat_hash(salt, nonce, mac, enc_data, protection_type):
+    print("\nFound hashcat hash!")
+    if protection_type == 0x2000:
+        versions = BITLOCKER_PASSWORD_HASH_VERSIONS
+    if protection_type == 0x800:
+        versions = BITLOCKER_RECOVERY_HASH_VERSIONS
```

Comment:
The comment at start of script says this won't extract recovery key VMKs, but then we do have this in code - what exactly does that mean? Will the script actually extract such data, or is this functionality in some way incomplete?